### ISSUE TITLE
How to trim mirrored part of FFT from CLFFT1D ?

### ISSUE THREAD

**Issue by Rares14324** [ARM-software/ComputeLibrary#791]
**Problem description:**
Hello I want to implement part of my processing chain which deals with discrete fourier transform using ACL. The result of applying DFT is mirrored at half the dimension and I don't need the other half. My problem is how should I eliminate it ? My DFT is a real to complex application so the end result has 2 channels. What I tried was CLReshapeLayer which doesn't work on on 2 channels tensors, CLSlice which also doesn't work on 2 channels Tensors. The following snippet illustrates my need (include paths need to be set)
```
void prototype_feat_extraction(int sizeFFt, int timeBins, const std::vector<float>& signalIn)
{
    //int overlapInSamples = sizeFFt * 3 / 4;
    sizeFFt = 512;
    timeBins = 65;
    CLTensor inputAudio = test::create_tensor<CLTensor>(TensorShape(sizeFFt, timeBins), DataType::F32);
    inputAudio.info()->set_data_layout(arm_compute::DataLayout::NHWC);
    CLTensor inputHannWindow = test::create_tensor<CLTensor>(TensorShape(sizeFFt,sizeFFt), DataType::F32);
    inputHannWindow.info()->set_data_layout(arm_compute::DataLayout::NHWC);
    CLLocallyConnectedMatrixMultiplyKernel multiplyWithHannWindow;
    CLFFT1D doFFT;
    CLSlice stripMirroredParts;
    FFT1DInfo defaultFFTSetting;
    CLTensor outputAudioWindowed = test::create_tensor<CLTensor>(TensorShape(sizeFFt, timeBins), DataType::F32);
    inputHannWindow.info()->set_data_layout(arm_compute::DataLayout::NHWC);
    multiplyWithHannWindow.configure(&inputAudio, &inputHannWindow, &outputAudioWindowed);
    CLTensor outputFFT1DComplex = test::create_tensor<CLTensor>(TensorShape(sizeFFt, timeBins), DataType::F32, 2);
    outputFFT1DComplex.info()->set_data_layout(arm_compute::DataLayout::NHWC);
    CLTensor outputFFT1DComplexHalf = test::create_tensor<CLTensor>(TensorShape(sizeFFt/2 + 1, timeBins), DataType::F32, 2); // only need half the data, the mirrored part needs to be discarded as it is redundant for my purposes
    outputFFT1DComplexHalf.info()->set_data_layout(arm_compute::DataLayout::NHWC);

    doFFT.configure(&outputAudioWindowed, &outputFFT1DComplex, defaultFFTSetting);
    stripMirroredParts.configure(&outputFFT1DComplex, &outputFFT1DComplexHalf, Coordinates(sizeFFt/2+1,0U), Coordinates(sizeFFt, timeBins));
// allocation and filling has been skipped
CLScheduler::get().enqueue(multiplyWithHannWindow, true);
    doFFT.run();
    stripMirroredParts.run();
}
```
Can this stripping be done with ACL somehow ?
Thank-you

**Comment by giuseros** [comment]
Hello @Rares14324 ,
You are right: CLSlice is what you are supposed to use, but for now it only supports non-complex data (i.e., channels=1). 

Complex operators support in ACL is very limited, since we introduced the FFT mainly to support fast(er) convolutions. 

Do you need to do more ACL manipulation after the stripping? In other words, is it only the CLSlice function that is stopping you?

All in all, I think that, depending on the urgency, you have the following options:
a) Mapping the tensor on the CPU, manually stripping it, and then copying back on the GPU. 
b) Modifying the CLSliceKernel yourself and pushing it here: https://review.mlplatform.org/#/c/ml/ComputeLibrary/+/270/. 
c) Waiting on us to add support for it in this or next release. Once the support is added we might point you at the development branch above, so that you don't have to wait the whole release cycle. 

Thanks,
Giuseppe

**Comment by Rares14324** [comment]
Hello @giuseros 
Thank-you for your answer.
 Yes I require more processing to be done after stripping half of the coefficients. I believe although support for complex is limited inside ACL a user can get by if we could extract the channels through something like CLChannelExtractKernel by passing Channel::C1, Channel::C0 to its arguments or possibly by making use of CLSplit and have it split the channels of the input tensor into separate tensors and from then most of ACL operations should be available again. It's just my opinion though.

With respect to the solutions provided a) for my purposes falls down because of the possible overhead of going from GPU->CPU->GPU and the fact that our current implementation has bits of it already in neon for better CPU usage. The reason why I investigated implementing it with ACL was for the benefit of executing on the GPU (where available). Regarding option b) I have looked into tailoring a kernel that both strips half the input and splits the channels but my knowledge of OpenCL is lacking and could take a while (looking at channel_extract.cl provides me with some inside to short-circuit the learning curve of OpenCL). I would very much welcome the addition of it inside the library (option c) ).  I don't mind cherry-picking commits for custom builds.

This is only a suggestion but it would be efficient to have an option inside the FFT1DInfo to specify when we only need half the coefficient to signal the algorithm not to compute them to save time. I don't know how hard that would be implementation-wise though.

So in short if you put this on the next features in release I would be thankful and happy to cherry-pick it. If time permits I will also try to implement it on my own but most likely it would be blunky and would only suit my specific need and will not be worth committing.

Rares

**Comment by giuseros** [comment]
Hi @Rares14324 ,
May I ask a bit more about what are you trying to do overall? We can plan a complex CLSlice, but if you need additional functions later on you will be stuck again. 

In other word, what is your use case, and what functionalities would you need in ACL to achieve that?

Thanks,
Giuseppe

**Comment by Rares14324** [comment]
Hello @giuseros 

The processing chain I want to implement for an audio application which extracts magnitude and phase from a short-term fourier transform on an audio signal. 
ACL provides the functions I need further along but they don't operate on complex data. This is why I suggested if you could possibly have CLChannelExtractKernel extract these channels to a new tensor ( by providing Channel::C1 as argument for example). The output tensor of the result would then be of 1 channel and all functions would work on it, CLSlice and the others (mostly PixelWiseOperationKernel operations is what I need).  I believe the effort would be less than implement all my required operations for complex tensors.

**Comment by giuseros** [comment]
Hi @Rares14324 ,
Thanks a lot for your reply. We will consider this use-case for future releases. Either we will start expanding our complex support or we will introduce the extraction function (or possibly both). 

Thank you,
Giuseppe
